File: B1/B1.cpp
// B1. Dots naturāls skaitis n. Atrast visus tādus naturālu skaitļu pārīšus a
// un b, kuri nepārsniedz n, lai a būtu  vienāds ar visu b dalītāju summu
// (izņemot b), un b būtu vienāds ar visu a dalītāju summu (izņemot a).
// Risinājumā izmantot funkciju, kas atpazīst pirmskaitļus.

#include <iostream>

using namespace std;

int divisorSum(int x)
{
    int sum = 1;

    for (int i = 2; i * i <= x; i++)
    {
        if (!(x % i))
        {
            if (i * i == x)
            {
                sum += i;
            }
            else
            {
                sum += i + (x / i);
            }
        }
    }

    return sum;
}

void outputPairsUnderN(int &n)
{
    for (int i = 1; i * 2 <= n; i++)
    {
        if (divisorSum(i) > n)
        {
            continue;
        }
        if (i == divisorSum(divisorSum(i)) && i <= divisorSum(i))
        {
            cout << i << " " << divisorSum(i) << endl;
        }
    }
}

int main()
{
    int n;
    cin >> n;

    outputPairsUnderN(n);

    return 0;
}

File: B1/B1.py
#  B1. Dots naturāls skaitis n. Atrast visus tādus naturālu skaitļu pārīšus a
#  un b, kuri nepārsniedz n, lai a būtu  vienāds ar visu b dalītāju summu
#  (izņemot b), un b būtu vienāds ar visu a dalītāju summu (izņemot a).
#  Risinājumā izmantot funkciju, kas atpazīst pirmskaitļus.

def divisorSum(x):
    sum = 1

    i = 2
    while (i * i <= x):
        if not (x % i):
            if i * i == x:
                sum += i
            else:
                sum += i + (x // i)
        i += 1

    return sum


def outputPairsUnderN(n):
    i = 1
    while (i <= n):
        if i == divisorSum(divisorSum(i)) and i >= divisorSum(i):
            print(f'{i} {divisorSum(i)}')
        i += 1


def main():
    n = int(input())

    outputPairsUnderN(n)


if __name__ == '__main__':
    main()
